ENH Adds n_features_to_select upper limit to SeqFeatureSelector

murata-yu opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Fixes #20137
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
Add `aborted_rate` parameter in SequentialFeatureSelector.
If aborted_rate is NOT None, Fitting is aborted if new_score is lower than old_score plus aborted_rate.

#### Any other comments?
I think it shoud make tests that aborted_rate is set.
but, I have no ideas for tests. please let me know if you have good ideas.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

murata-yu committed:
Fix sklearn.feature_selection.SequentialFeatureSelector Select features as long as score gets better. #20137

Add `censored_rate` parameter in SequentialFeatureSelector.
If censored_rate is NOT None, Fitting is aborted if new_score is lower than old_score*(1+censoring_rate).

murata-yu committed:
Fix sklearn.feature_selection.SequentialFeatureSelector Select features as long as score gets better. #20137

Add `censored_rate` parameter in SequentialFeatureSelector.
If censored_rate is NOT None, Fitting is aborted if new_score is lower than old_score*(1+censoring_rate).

murata-yu committed:
Merge branch 'my_feature' of https://github.com/murata-yu/scikit-learn into my_feature

murata-yu committed:
fix docstring

A more appropriate explanation

murata-yu committed:
fix iptimal name

murata-yu committed:
fix abort check

A more simplify

murata-yu committed:
add test parameter `aborted_rate`

[TODO] make tests for `aborted_rate` is not None

ogrisel commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -36,6 +36,11 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin,
         to select. If float between 0 and 1, it is the fraction of features to
         select.
 
+    aborted_rate : float, default=None
```
Please rename to `tol`.

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/commit/44300288cc81e213203febe6eddaf928fcbb48f1#diff-521200bea206f673d4e933a33304f606f951d97edd7999b4d6b019cca79a81a0L39

ogrisel commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -36,6 +36,11 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin,
         select.
 
+    aborted_rate : float, default=None
+	If the score is not increasing above ths rate, stop adding new features
+	even if n_features_to_select has not been reached. If None (default)
+	all n_features_to_select will be selected.
```
Maybe make it more explicit that this a threshold on the increment:

If the score is not incremented by at least `tol` between two consecutive feature addition or removal, stop adding or removing ...

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/44300288cc81e213203febe6eddaf928fcbb48f1

Maybe I don't fully understand the code, but can the features be reduced?
Does that mean reducing from the original features?

ogrisel replied:
Yes, look at the docstring for `direction="backward"`.

murata-yu replied:
thanks, I understand !

ogrisel commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -175,18 +181,34 @@ def fit(self, X, y):
-            current_mask = ~current_mask
+        if self.aborted_rate is None:
+            for _ in range(n_iterations):
+                new_feature_idx, new_score = self._get_best_new_feature_score(cloned_estimator, X,
+                                                             y, current_mask)
```
code style (PEP8):

```suggestion
                new_feature_idx, new_score = self._get_best_new_feature_score(
                    cloned_estimator, X, y, current_mask
                )
```

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/2922dbba5e33e7b0e6990ba71cc33372307ca9a0

I tried using `autopep8`. It's so useful!
however, The position of the parentheses is different from the code @ogrisel  suggested ...
Should I fix it?

ogrisel commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -175,18 +181,34 @@ def fit(self, X, y):
+        else:
+            old_score = 0
+            for _ in range(n_iterations):
+                new_feature_idx, new_score = self._get_best_new_feature_score(cloned_estimator, X,
+                                                             y, current_mask)
```
```suggestion
                new_feature_idx, new_score = self._get_best_new_feature_score(
                    cloned_estimator, X, y, current_mask
                )
```

ogrisel reviewed (commented):
First round of comments:

murata-yu committed:
lint docstring

NicolasHug commented:
Thanks for the fast PR @murata-yu  but please note that for now, we're still discussing alternative solutions in https://github.com/scikit-learn/scikit-learn/issues/20137 so the requirements might change

murata-yu committed:
rename aborted_rate to tol

ogrisel commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -39,9 +40,11 @@ def test_n_features_to_select(direction, n_features_to_select):
+        assert sfs.n_features_to_select_ == n_features_to_select
+        assert sfs.transform(X).shape[1] == n_features_to_select
 
 
```
if `tol=1e-3` and `n_features_to_select=X.shape[1]` you can check that the number of selected features is lower than `X.shape[1]`.

You can also check that the cross-validation score of a model fitted on a strict random subset of the selected feature is lower than the mean cross-validation score of the final model by more than `tol`.

Maybe I would put this in a dedicated new test named `test_stopping_criterion`.

ogrisel replied:
And you can write a similar (dual) test `test_backward_stopping_criterion` for the backward strategy.

ogrisel replied:
Maybe it would be easier to write 2 independent tests for the 2 directions.

In both tests, we should check that when using `tol=1e-3` and `n_features_to_select=X.shape[1]` the final number of selected features `selector.support_.sum()` is strictly smaller than `X.shape[1]`.

Furthermore the cross-validation score of the model without feature selection should be worse than the CV score of the model with feature selection (without using `tol` in this comparison).

Then for the forward test, add the following assertion:

- select a new features at random among those not currently selected in `selector.support_`, build a new version of the data that includes all the features in `selector.support_` + this newly selected feature and check that the cross-validation score of the model trained on this new dataset variant is lower than the model with the selected forward selected features or at least does not improve by more than the tol margin.

Then for the backward direction do the opposite check: instead of adding a new feature to `selector.support_`, try to remove one of those selected features at random and check that the cross-validation score is either decreasing or not improving by more than the tol margin.

murata-yu replied:
added test for score comparison on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/05e86300a864d1e8f5cd439630143fa2d1d0ecc5, https://github.com/scikit-learn/scikit-learn/pull/20145/commits/2f12708e2fe0432dddb0df2d224f02a1ef5d6dc1 and https://github.com/scikit-learn/scikit-learn/pull/20145/commits/f047e05c91abd2c5bb2dd2633ab203b8f3052c26.

And, added test for checking num of features on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/2640d7873cbf6400f3bf2d7e1d6d31f165e4f2c4.

ogrisel commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -175,18 +181,34 @@ def fit(self, X, y):
+                new_feature_idx, new_score = self._get_best_new_feature_score(cloned_estimator, X,
+                                                             y, current_mask)
+                if new_score - self.aborted_rate < old_score:
+                    break
```
I would find it easier to follow with:

```suggestion
                if (new_score - old_score) < self.tol:
                    break
```

ogrisel commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -175,18 +181,34 @@ def fit(self, X, y):
+            if self.direction == 'backward':
+                current_mask = ~current_mask
+        else:
+            old_score = 0
```
To make sure that the first iteration is always accepted, whatever the value of `self.scoring`.

```suggestion
            old_score = -np.inf
```

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/e47eea7c256df2355e2143decb875ac869076b56

ogrisel commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -175,18 +181,34 @@ def fit(self, X, y):
+            for _ in range(n_iterations):
+                new_feature_idx, new_score = self._get_best_new_feature_score(cloned_estimator, X,
+                                                             y, current_mask)
+                if new_score - self.tol < old_score:
```
If should be possible to move the `if self.tol is None:` condition inside the loop, here to avoid code duplication:

```suggestion
                if self.tol is not None and new_score - self.tol < old_score:
```

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/1c614a8f4246f405a5868133f4cb9d57fd069d0b

murata-yu committed:
fix docstring

more explicit

murata-yu committed:
formatt by autopep8

murata-yu committed:
Add test for tol

Backward test is failing.
May need to change the test content

murata-yu committed:
Fix initial value of old_score

To make sure that the first iteration is always accepted.

murata-yu committed:
Avoid code dupulication

ogrisel commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -39,11 +40,44 @@ def test_n_features_to_select(direction, n_features_to_select):
+def test_stopping_criterion(direction):
+    # Make sure n_features_to_select is respected
+
+    X, y = make_regression(n_features=10)
```
Maybe add a large fraction of non-informative features with `n_features=50` and `n_informative=10` to make the impact of feature selection more visible.

ogrisel replied:
Also please pass `random_state=0` to make this test deterministic.

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/b5a332eb2f87b74ec9a97f13a6fdb116372a0fa2

ogrisel commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -39,11 +40,44 @@ def test_n_features_to_select(direction, n_features_to_select):
+        sfs.transform(X),
+        y,
+        cv=2).mean()
+    def_cv_score = cross_val_score(LinearRegression(), X, y, cv=2).mean()
```
Let's use more explicit variable names:

```suggestion
    all_features_cv_score = cross_val_score(LinearRegression(), X, y, cv=2).mean()
```

ogrisel replied:
This change will require additional updates in the test.

ogrisel reviewed (commented):
Thanks @murata-yu, here are more details about what I had in mind for the test.

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

murata-yu committed:
increase n_features to 50

add n_infomative and random_state

murata-yu committed:
fix typo

murata-yu committed:
fix n_iterations is not inverted when tol is enabled

murata-yu committed:
fix typo

murata-yu committed:
improve test_stopping_criterion

ogrisel reviewed (commented):
This PR is looking good. What remains to be done is implement the deprecation strategy presented in https://github.com/scikit-learn/scikit-learn/issues/20137#issuecomment-851977247 and update the test accordingly to use the "auto" mode.

The docstrings will for `n_features_to_select` and `tol` also need to be updated.

And we will need an entry to document this new feature and the deprecation in `doc/whats_new/v1.0.rst`.

Once this is done, feel free to rename the PR to remove the `[WIP]` marker to attract reviewers' attention.

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

murata-yu committed:
update docstring to apply auto mode

murata-yu committed:
update to apply auto mode

murata-yu committed:
fix test_n_features_to_select to apply auto mode

murata-yu committed:
fix test_stopping_criterion to apply auto mode

remove selected feature comparison because it implemented in

murata-yu committed:
fix bad evaluation

murata-yu committed:
fix `n_features_to_select_` to return the num of selected feature when
auto mode

add test if `n_features_to_select_` is correct

murata-yu committed:
add tests to remove features

murata-yu committed:
fix bad evaluation

murata-yu committed:
add whats_new

murata-yu committed:
fix linting

murata-yu committed:
undo tol to 1e-3

murata-yu committed:
add test to compare feature selection or not

norbusan commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -30,11 +30,17 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin,
-        select.
+    n_features_to_select : int or float, default='auto'
+        The number of features to select. If `'auto'`, features are selected
+        untill score improvement does not exceed tol. If integer, the parameter
```
Should be `until` not `untill`

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/3a6ebb4dc22d96edd81ee63120f42bfbf8b899be

norbusan commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -30,11 +30,17 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin,
+        it is the fraction of features to select.
+
+    tol : float, default=0.0
+        If the score is not incremented by at least tol between two consecutive
```
backticks around `tol`

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/3a6ebb4dc22d96edd81ee63120f42bfbf8b899be

norbusan commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -30,11 +30,17 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin,
+    tol : float, default=0.0
+        If the score is not incremented by at least tol between two consecutive
+        feature addition or removal, stop adding or removing even if
+        n_features_to_select has not been reached. tol is enabled only when
```
Backticks around `n_features_to_select` and `tol`

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/3a6ebb4dc22d96edd81ee63120f42bfbf8b899be

norbusan commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -145,8 +153,8 @@ def fit(self, X, y):
                      "number of features, or a float in (0, 1] "
                      "representing a percentage of features to "
                      f"select. Got {self.n_features_to_select}")
-        if self.n_features_to_select is None:
```
Shouldn't we keep the `None` case for now for backwards compatibility?

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/c6c4b2eb466c77e8ba7ad0f8cdf658ea8ab77449.

Simply, convert 'auto' when  `self.n_features_to_select` is `None`.
Should I send a message? 
(e.g. self.n_features_to_select=None is deprecated)

murata-yu replied:
Fixed not to change self.n_features_to_select to 'auto' just by issuing a warning when `n_features_to_select` is 'warn' on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/83bd9a064557e3a5fd995c720f1bce09560f4d58 because assertion was occured in https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/utils/estimator_checks.py#L2731

murata-yu committed:
fix typo in docstring

murata-yu committed:
add backwards compability

murata-yu committed:
fix linting

murata-yu committed:
remove None pattern

murata-yu committed:
fix lint in docstring

murata-yu committed:
fix lint in docstring

NicolasHug commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -30,11 +30,17 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin,
-        selected. If integer, the parameter is the absolute number of features
-        to select. If float between 0 and 1, it is the fraction of features to
-        select.
+    n_features_to_select : int or float, default='auto'
```
```suggestion
    n_features_to_select : int or float, default='warn'
```

NicolasHug commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -30,11 +30,17 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin,
+        If the score is not incremented by at least `tol` between two
+        consecutive feature addition or removal, stop adding or removing even
+        if `n_features_to_select` has not been reached. `tol` is enabled only
+        when `n_features_to_select` is `'auto'`.
```
Below we should add

`.. versionadded:: 1.0`

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/9b11fbbcfb1b443063a25940f4f8d3f5a4c6951e

NicolasHug commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -30,11 +30,17 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin,
+        The number of features to select. If `'auto'`, features are selected
+        until score improvement does not exceed `tol`. If integer, the
+        parameter is the absolute number of features to select. If float
+        between 0 and 1, it is the fraction of features to select.
```
Below we should add

```
.. versionadded:: 1.0
    The default changed from None to 'warn' in 1.0 and will become 'auto' in 1.2.
    To keep the same behaviour as with `None`, you should manually set
    `n_features_to_select='auto'` and set `tol=None`
```

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/9b11fbbcfb1b443063a25940f4f8d3f5a4c6951e

NicolasHug commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -30,11 +30,17 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin,
-        to select. If float between 0 and 1, it is the fraction of features to
-        select.
+    n_features_to_select : int or float, default='auto'
+        The number of features to select. If `'auto'`, features are selected
```
We should be clearer here, with a bullet list:

```
If 'auto', the behaviour depends on the `tol` parameter:

- if `tol` is not None, then features are selected until the score improvement does not exceed `tol`.
- otherwise, half of the features are selected.
```

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/9b11fbbcfb1b443063a25940f4f8d3f5a4c6951e

NicolasHug commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -145,8 +152,13 @@ def fit(self, X, y):
         if self.n_features_to_select is None:
-            self.n_features_to_select_ = n_features // 2
+            # for backwards compability
+            self.n_features_to_select = 'auto'
```
This should be 'warn'

And then if n_features_to_select is 'warn', we should add a warning like

>  Leaving n_features_to_select to None is deprecated in 1.0 and will become 'auto' in 1.2.
    To keep the same behaviour as with `None` (i.e. select half of the features) and avoid this warning, you should manually set
    `n_features_to_select='auto'` and set `tol=None`.

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/9b11fbbcfb1b443063a25940f4f8d3f5a4c6951e

NicolasHug commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -172,21 +184,32 @@ def fit(self, X, y):
+            new_feature_idx, new_score = self._get_best_new_feature_score(
+                cloned_estimator, X, y, current_mask)
+            if self.n_features_to_select == 'auto' and (
+                    new_score - old_score) < self.tol:
```
We'll probably need to check that self.tol is not None first

murata-yu replied:
fix it on https://github.com/scikit-learn/scikit-learn/pull/20145/commits/9b11fbbcfb1b443063a25940f4f8d3f5a4c6951e

NicolasHug reviewed (commented):
Hi @murata-yu , thanks for the PR! I made a quick pass.

I believe we have decided to go with this deprecation strategy now: https://github.com/scikit-learn/scikit-learn/issues/20137#issuecomment-851977247

(I assume @ogrisel 's +1 means he's happy with it)

so we will need a few changes to the current implementation

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Nicolas Hug <contact@nicolas-hug.com>

murata-yu committed:
fix default params for `n_featues_to_select` and `tol`

- fix default params for `n_featues_to_select` and `tol` in docstring
- add versionadded in docstring
- add warn if using `n_features_to_select`'s defaut param
- add `tol` check in stopping criterion
- add test when `tol` is None or not

murata-yu committed:
Merge branch 'my_feature' of https://github.com/murata-yu/scikit-learn into my_feature

murata-yu committed:
fix lint

murata-yu committed:
fix lint

murata-yu committed:
Fixed not to change `self.n_features_to_select` to 'auto' just by issuing a warning when `n_features_to_select` is 'warn'

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/murata-yu/20145

glemaitre committed:
remove None in test

glemaitre commented:
I resolve the conflict. I will make a review now.

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -29,11 +31,29 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator
+        If integer, the parameter is the absolute number of features to select.
+        If float between 0 and 1, it is the fraction of features to select.
+
+        .. versionadded:: 1.0
+            The default changed from None to 'warn' in 1.0 and will become
+            'auto' in 1.2. To keep the same behaviour as with `None`, you
+            should manually set `n_features_to_select='auto'` and
+            set `tol=None`
```
```suggestion
        .. versionadded:: 1.0
           The option `"auto"` was added in version 1.0.
           
        .. deprecation:: 1.0
           The default changed from `None` to `"warn"` in 1.0 and will become
           `"auto"` in 1.2. To keep the same behaviour as with `None`, you
           should manually set `n_features_to_select="auto"` and
           set `tol=None`.
```

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -29,11 +31,29 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator
-        selected. If integer, the parameter is the absolute number of features
-        to select. If float between 0 and 1, it is the fraction of features to
-        select.
+    n_features_to_select : int or float, default='warn'
```
```suggestion
    n_features_to_select : "auto", int or float, default='warn'
```

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -29,11 +31,29 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator
-        to select. If float between 0 and 1, it is the fraction of features to
-        select.
+    n_features_to_select : int or float, default='warn'
+        If 'auto', the behaviour depends on the `tol` parameter:
```
```suggestion
        If `"auto"`, the behaviour depends on the `tol` parameter:
```

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -29,11 +31,29 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator
+    n_features_to_select : int or float, default='warn'
+        If 'auto', the behaviour depends on the `tol` parameter:
+
+        - if `tol` is not None, then features are selected until the score
```
```suggestion
        - if `tol` is not `None`, then features are selected until the score
```

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -29,11 +31,29 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator
+            should manually set `n_features_to_select='auto'` and
+            set `tol=None`
+
+    tol : float, default=0.0
```
```suggestion
    tol : float, default=None
```

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -29,11 +31,29 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator
+
+    tol : float, default=0.0
+        If the score is not incremented by at least `tol` between two
+        consecutive feature addition or removal, stop adding or removing even
```
```suggestion
        consecutive feature additions or removals, stop adding or removing even
```

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -29,11 +31,29 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator
+        If the score is not incremented by at least `tol` between two
+        consecutive feature addition or removal, stop adding or removing even
+        if `n_features_to_select` has not been reached. `tol` is enabled only
+        when `n_features_to_select` is `'auto'`.
```
```suggestion
        when `n_features_to_select` is `"auto"`.
```

glemaitre replied:
I assume that we should as well document the `None` option. From the deprecation warning, I was under the impression that it allows for backward compatibility.

murata-yu replied:
add docstring about None option with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/744c2358e0ab4b1a79fbfd9e55cc134ab7187cce

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -125,11 +146,24 @@ def __init__(
+                "(i.e. select half of the features) and avoid "
+                "this warning, you should manually set "
+                "n_features_to_select='auto' and set tol=None.",
+                PendingDeprecationWarning,
```
Usually, we are using a `FutureWarning`. Is there a reason for not using it here?

murata-yu replied:
There is no particular reason, so I fixed it with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/0d1b5f3fde10e1f0ee5bb664fbdd740bd7b9f03d

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -190,22 +227,33 @@ def fit(self, X, y):
 
-    def _get_best_new_feature(self, estimator, X, y, current_mask):
+    def _get_best_new_feature_score(self, estimator, X, y, current_mask):
         # Return the best new feature to add to the current_mask, i.e. return
```
We should also update the docstring of this method. Could you make a proper docstring instead of the current comment?

murata-yu replied:
update docstring with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/3e098b0debc7dc64ffcad3fbf3919a926d9bfeb2

glemaitre commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -41,21 +42,94 @@ def test_n_features_to_select(direction, n_features_to_select):
+@pytest.mark.parametrize("direction", ("forward", "backward"))
+@pytest.mark.parametrize("tol", (1e-3, None))
+def test_n_features_to_auto_select(direction, tol):
+    # Make sure n_features_to_select is respected
```
```suggestion
    """Check the behaviour of `n_features_to_select="auto"` with different 
    values for the parameter `tol`.
    """
```

glemaitre commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -41,21 +42,94 @@ def test_n_features_to_select(direction, n_features_to_select):
 
+@pytest.mark.parametrize("direction", ("forward", "backward"))
+@pytest.mark.parametrize("tol", (1e-3, None))
+def test_n_features_to_auto_select(direction, tol):
```
```suggestion
def test_n_features_to_select_auto(direction, tol):
```

glemaitre commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -41,21 +42,94 @@ def test_n_features_to_select(direction, n_features_to_select):
+
+
+@pytest.mark.parametrize("direction", ("forward", "backward"))
+def test_stopping_criterion(direction):
```
```suggestion
def test_n_features_to_select_stopping_criterion(direction):
```

glemaitre commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -41,21 +42,94 @@ def test_n_features_to_select(direction, n_features_to_select):
+
+@pytest.mark.parametrize("direction", ("forward", "backward"))
+def test_stopping_criterion(direction):
+    # Make sure n_features_to_select is respected
```
```suggestion
    """Check the behaviour stopping criterion for feature selection
    depending on the values of `n_features_to_select` and `tol`.
    """
```

glemaitre reviewed (commented):
Since we are raising a warning for the deprecation, we should force `n_features_to_select="auto"` and `tol=None` in the tests. We should as well add a test that checks that we raise the deprecation warning.

glemaitre commented on doc/whats_new/v1.0.rst:
```diff
@@ -335,6 +335,11 @@ Changelog
 
+- |Feature| Added auto mode to :class:`feature_selection.SequentialFeatureSelection`.
+  If the argument `n_features_to_select` is `'auto'`, select features until
+  the score improvement does not exceed the argument `tol`.
```
We should as well the deprecation cycle regarding the default value of `n_features_to_select`

murata-yu replied:
fix it with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/f843d3772b846c8596462a8fbdc856c95f888109

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
fix lint

murata-yu committed:
add a test that checks that we raise the deprecation warning

murata-yu committed:
fix lint

murata-yu committed:
change warning to FutureWarning

murata-yu committed:
update the docstring of _get_best_new_feature_score

murata-yu committed:
add documents of the deprecation cycle regarding the default value of n_features_to_select

murata-yu committed:
add to docstring that  has been disabled

murata-yu commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -41,21 +42,98 @@ def test_n_features_to_select(direction, n_features_to_select):
 
+@pytest.mark.parametrize("direction", ("forward", "backward"))
+@pytest.mark.parametrize("tol", (1e-3, None))
+def test_n_features_to_select_auto(direction, tol):
```
> Since we are raising a warning for the deprecation, we should force n_features_to_select="auto" and tol=None in the tests. 

There are tests that explicitly takes `"auto"` as an argument, is that insufficient?

glemaitre replied:
yes it would be enough in this case

murata-yu replied:
force n_features_to_select="auto" with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/8f2b71918369f342b5a758800a09e1f5d709dedf

murata-yu commented:
> We should as well add a test that checks that we raise the deprecation warning.

I add warning test with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/66139a397a525c2e3ca9847871f1dd7fe30932f4

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -29,11 +31,33 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator
+
+        .. deprecation:: 1.0
+           The default changed from `None` to `"warn"` in 1.0 and will become
+           `"auto"` in 1.2. `None` has been disabled. `'warn'` is deprecated
```
```suggestion
           `"auto"` in 1.2. `None` has been disabled. `"warn"` is deprecated
```

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -125,11 +150,24 @@ def __init__(
+        if self.n_features_to_select == "warn":
+            # for backwards compability
+            warnings.warn(
+                "Leaving n_features_to_select to "
```
```suggestion
                "Leaving `n_features_to_select` to "
```

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -125,11 +150,24 @@ def __init__(
+                "in 1.2. To keep the same behaviour as with None "
+                "(i.e. select half of the features) and avoid "
+                "this warning, you should manually set "
+                "n_features_to_select='auto' and set tol=None.",
```
```suggestion
                "`n_features_to_select='auto'` and set `tol=None`.",
```

glemaitre commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -145,3 +223,8 @@ def test_pipeline_support():
+
+def test_raise_deprecation_warning():
+    with pytest.warns(FutureWarning):
+        SequentialFeatureSelector(LinearRegression())
```
The warning should happen only when calling fit. This test should fail indeed.

murata-yu replied:
warning is called from `__init__`

https://github.com/scikit-learn/scikit-learn/pull/20145/files#diff-521200bea206f673d4e933a33304f606f951d97edd7999b4d6b019cca79a81a0R159-R169

glemaitre replied:
Right, so I badly reviewed :). You should move this warning to `fit`. We always do parameter checking at `fit`. `__init__` is only to set the parameter as attributes.

murata-yu replied:
thanks, I understand !
move warnings from __init__ to fit and fix warning description with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/05ffbdaab84b8d18f246497e65b0b7d306d681ad

glemaitre commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -145,3 +223,8 @@ def test_pipeline_support():
+
+
+def test_raise_deprecation_warning():
+    with pytest.warns(FutureWarning):
```
It will be better to match a partial string to be sure that we raise the expected warning:

```suggestion
    warn_msg = "Leaving ...."
    with pytest.warns(FutureWarning, match=warn_msg):
```

glemaitre commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -145,3 +223,8 @@ def test_pipeline_support():
     pipe.transform(X)
+
+
+def test_raise_deprecation_warning():
```
```suggestion
# FIXME : to be removed in 1.2
def test_raise_deprecation_warning():
    """Check that we raise a deprecation warning with `n_features_to_select`."""
```

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
move warnings from __init__ to fit

murata-yu committed:
Merge branch 'my_feature' of https://github.com/murata-yu/scikit-learn into my_feature

murata-yu committed:
force n_features_to_select='auto' to avoid warnings

murata-yu committed:
move warnings from __init__ to fit

glemaitre commented:
The remaining failure is an easy one to solve. You need to set the parameters in the `test_fit_docstring_attributes` (l. 264) of the file `sklearn/test/test_docstring_parameters.py` with the following line:

```python
    # FIXME: TO BE REMOVED for 1.2 (avoid FutureWarning)
    if Estimator.__name__ == "SequentialFeatureSelector":
        est.set_params(n_features_to_select="auto")
```

It will avoid to raise the `FutureWarning`.

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -144,6 +170,19 @@ def fit(self, X, y):
         -------
         self : object
         """
+        if self.n_features_to_select == "warn":
```
```suggestion
        # FIXME: to be removed in 1.2
        if self.n_features_to_select == "warn":
```

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -144,6 +170,19 @@ def fit(self, X, y):
+        if self.n_features_to_select == "warn":
+            # for backwards compability
+            warnings.warn(
+                "Leaving n_features_to_select to "
```
```suggestion
                "Leaving `n_features_to_select` to "
```

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

glemaitre commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -144,6 +170,19 @@ def fit(self, X, y):
+                "in 1.2. To keep the same behaviour as with None "
+                "(i.e. select half of the features) and avoid "
+                "this warning, you should manually set "
+                "n_features_to_select='auto' and set tol=None "
```
```suggestion
                "`n_features_to_select='auto'` and set tol=None "
```

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

glemaitre commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -136,12 +222,25 @@ def test_pipeline_support():
+    n_samples, n_features = 50, 3
+    X, y = make_regression(n_samples, n_features, random_state=0)
+
+    warn_msg = "Leaving ...."
```
It would be better to put the beginning of the string indeed

```suggestion
    warn_msg = "Leaving n_features_to_select to None is deprecated"
```

glemaitre reviewed (approved):
Only minor comment. LGTM after the quick fixes.
Thanks @murata-yu. We will need a second review.
Probably @ogrisel or @thomasjpfan might be able to jump in

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

murata-yu committed:
add param to test_fit_docstring_attributes

murata-yu committed:
Merge branch 'my_feature' of https://github.com/murata-yu/scikit-learn into my_feature

murata-yu committed:
fix typo

glemaitre commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -136,12 +222,25 @@ def test_pipeline_support():
+    n_samples, n_features = 50, 3
+    X, y = make_regression(n_samples, n_features, random_state=0)
+
+    warn_msg = "Leaving n_features_to_select to None is deprecated"
```
```suggestion
    warn_msg = "Leaving `n_features_to_select` to None is deprecated"
```

glemaitre commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -136,12 +222,25 @@ def test_pipeline_support():
+
+# FIXME : to be removed in 1.2
+def test_raise_deprecation_warning():
+    """Check that we raise a deprecation warning with `n_features_to_select`."""
```
```suggestion
    """Check that we raise a FutureWarning with `n_features_to_select`."""
```

glemaitre reviewed (approved):
I will merge these 2 small fixes that should make the CIs pass

glemaitre committed:
Apply suggestions from code review

thomasjpfan commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -23,13 +24,15 @@ def test_bad_n_features_to_select(n_features_to_select):
 
 @pytest.mark.parametrize("direction", ("forward", "backward"))
-@pytest.mark.parametrize("n_features_to_select", (1, 5, 9, None))
+@pytest.mark.parametrize("n_features_to_select", (1, 5, 9))
```
I think we can add "auto" here to test the `tol=None` + `"auto"` case. This way we can simplify the `test_n_features_to_select_auto` to test check for `tol=1e-3` and `'auto'`.

murata-yu replied:
simplify test with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/649fc31e9725970ab82845de8b4774bc1894472f

I'm not confident in this fix.
Does this fix meet @thomasjpfan needs?

thomasjpfan reviewed (commented):
Thank you for the PR @murata-yu !

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
fix test to deterministic

murata-yu committed:
add more detailed docstring

murata-yu committed:
clarify n_features and random_state

murata-yu committed:
simplify test_n_features_to_select_auto

murata-yu commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -41,21 +44,98 @@ def test_n_features_to_select(direction, n_features_to_select):
+    """Check the behaviour stopping criterion for feature selection
+    depending on the values of `n_features_to_select` and `tol`.
+    """
+
```
fix with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/2c61384d6e9a976d46edda3be5bc13cfbd5c35ca

murata-yu commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -41,21 +44,98 @@ def test_n_features_to_select(direction, n_features_to_select):
+
+@pytest.mark.parametrize("direction", ("forward", "backward"))
+def test_n_features_to_select_stopping_criterion(direction):
+    """Check the behaviour stopping criterion for feature selection
+    depending on the values of `n_features_to_select` and `tol`.
```
add docstring with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/76d33527c2808abbb65180438c66bfbbcc68b8f9

murata-yu commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -41,21 +44,98 @@ def test_n_features_to_select(direction, n_features_to_select):
+    values for the parameter `tol`.
+    """
+
+    X, y = make_regression(n_features=10)
```
clarify n_features with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/5327b6a022b5556ee9b3fce7fa2c7aa2964cf35c

thomasjpfan commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -23,39 +24,131 @@ def test_bad_n_features_to_select(n_features_to_select):
 
 @pytest.mark.parametrize("direction", ("forward", "backward"))
-@pytest.mark.parametrize("n_features_to_select", (1, 5, 9, None))
+@pytest.mark.parametrize("n_features_to_select", (1, 5, 9, "auto"))
```
Looking at this again, we need to continue to test `n_features_to_select=None` to make sure it works.

We can use a `@pytest.mark.filterwarnings("ignore:Leaving `n_features_to_select` to ")` to ignore the warning.

murata-yu replied:
add `None` with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/0dd1499a646bcfa47fadf2d3ca2db9e8329e54ee

thomasjpfan commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -105,7 +198,9 @@ def test_sparse_support():
     X, y = make_regression(n_features=10)
     X = scipy.sparse.csr_matrix(X)
-    sfs = SequentialFeatureSelector(LinearRegression(), cv=2)
+    sfs = SequentialFeatureSelector(
+        LinearRegression(), n_features_to_select="auto", cv=2
+    )
```
We still need to test `None` here. We can do this by parameterizing `test_sparse_support` to go through `None` and `'auto'`

```python
# TODO: Remove test for n_features_to_select=None in 1.2
@pytest.mark.filterwarnings("ignore:Leaving `n_features_to_select` to ")
@pytest.mark.parametrize("n_features_to_select", ["auto", None])
def test_sparse_support(n_features_to_select):
        ...
        sfs = SequentialFeatureSelector(
            LinearRegression(), n_features_to_select=n_features_to_select, cv=2
        )
```

murata-yu replied:
add `None` with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/0dd1499a646bcfa47fadf2d3ca2db9e8329e54ee

thomasjpfan commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -23,39 +24,131 @@ def test_bad_n_features_to_select(n_features_to_select):
         (0.1, 1),
         (1.0, 10),
         (0.5, 5),
-        (None, 5),  # just to make sure .5 is equivalent to passing None
```
`None` still needs to be tested.

murata-yu replied:
add `None` with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/b8a16a26ec05149d45cc22845570c05ab7408be3

thomasjpfan reviewed (commented):
Thanks for the update @murata-yu !

Looking at this more closely, we still need to make sure that `n_features_to_select=None` still works and continue to test it.

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
move the definition of is_auto_select outside of the loop

murata-yu committed:
Fixed to raise a warning even with None

murata-yu committed:
Add None to the parameter

murata-yu commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -190,25 +233,38 @@ def fit(self, X, y):
             )
+            is_auto_select = (
+                self.tol is not None and self.n_features_to_select == "auto"
+            )
```
move the definition with https://github.com/scikit-learn/scikit-learn/pull/20145/commits/37e28ee0b4732324c591536041d11156aff0942c

murata-yu committed:
Add None to the parameter

thomasjpfan commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -136,12 +240,25 @@ def test_pipeline_support():
     pipe.transform(X)
+
+
+# FIXME : to be removed in 1.2
```
```suggestion
# FIXME : to be removed in 1.3
```

thomasjpfan commented on sklearn/tests/test_docstring_parameters.py:
```diff
@@ -261,6 +261,10 @@ def test_fit_docstring_attributes(name, Estimator):
     if Estimator.__name__ == "TSNE":
         est.set_params(learning_rate=200.0, init="random")
 
+    # FIXME: TO BE REMOVED for 1.2 (avoid FutureWarning)
```
```suggestion
    # FIXME: TO BE REMOVED for 1.3 (avoid FutureWarning)
```

thomasjpfan commented on sklearn/feature_selection/tests/test_sequential.py:
```diff
@@ -100,12 +195,17 @@ def test_sanity(seed, direction, n_features_to_select, expected_selected_feature
 
 
-def test_sparse_support():
+# TODO: Remove test for n_features_to_select=None in 1.2
```
```suggestion
# TODO: Remove test for n_features_to_select=None in 1.3
```

thomasjpfan commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -155,16 +195,20 @@ def fit(self, X, y):
         )
         n_features = X.shape[1]
 
+        # FIXME: to be fixed in 1.2
```
```suggestion
        # FIXME: to be fixed in 1.3
```

thomasjpfan reviewed (commented):
Thank you for your patience with updating this PR.

Most of the comments has to do with moving the feature to 1.1. Otherwise, I think this PR is good to go.

murata-yu committed:
Update doc/whats_new/v1.0.rst

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/tests/test_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/tests/test_docstring_parameters.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

thomasjpfan commented on doc/whats_new/v1.0.rst:
```diff
@@ -335,6 +335,14 @@ Changelog
+  If the argument `n_features_to_select` is `'auto'`, select features until
+  the score improvement does not exceed the argument `tol`.
+  The default value of `n_features_to_select` changed from `None` to `‘warn’` in 1.1
+  and will become `'auto'` in 1.3. `None` has been disabled. `'warn'` is deprecated
+  in 1.0 and will be disabled in 1.3.
```
```suggestion
  and will become `'auto'` in 1.3. `None` and `'warn'` will be removed in 1.3.
```

thomasjpfan commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -29,11 +31,33 @@ class SequentialFeatureSelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator
+
+        .. deprecated:: 1.1
+           The default changed from `None` to `"warn"` in 1.1 and will become
+           `"auto"` in 1.3. `None` has been disabled. `"warn"` is deprecated
+           in 1.1 and will be disabled in 1.3. To keep the same behaviour
+           as with `None`, you should manually set `n_features_to_select="auto"`
+           and set `tol=None`.
```
```suggestion
           `"auto"` in 1.3. `None` and `'warn'` will be removed in 1.3.
           To keep the same behaviour as `None`, set 
           `n_features_to_select="auto" and `tol=None`.
```

thomasjpfan commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -144,6 +170,20 @@ def fit(self, X, y):
         -------
         self : object
         """
+        # FIXME: to be removed in 1.2
```
This still needs to be addressed

thomasjpfan commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -144,6 +170,20 @@ def fit(self, X, y):
+            # for backwards compability
+            warnings.warn(
+                "Leaving `n_features_to_select` to "
+                "None is deprecated in 1.0 and will become 'auto' "
+                "in 1.2. To keep the same behaviour as with None "
```
Same here

thomasjpfan commented on sklearn/feature_selection/_sequential.py:
```diff
@@ -155,16 +195,22 @@ def fit(self, X, y):
         )
-        if self.n_features_to_select is None:
-            self.n_features_to_select_ = n_features // 2
+        if self.n_features_to_select in ("auto", "warn", None):
```
In the edge case where `tol is not None` and `n_features_to_select=None`, we need to include the same code path as before:

```python
if self.n_features_to_select in ("warn", None):
    if self.tol is not None:
		raise ValueError("tol is only enabled if `n_features_to_select='auto'`")
    self.n_features_to_select_ = n_features // 2
elif self.n_features_to_select == "auto":
```

murata-yu replied:
fixed to be addressed until 1.3 https://github.com/scikit-learn/scikit-learn/pull/20145/commits/1ba4b4cbd2f5a7d0556114d06e734821ff70080c

fixed to include the same code path https://github.com/scikit-learn/scikit-learn/pull/20145/commits/23f722c686af450b8664ae53cdf59a7fed76d396

thomasjpfan reviewed (commented):
Sorry for the delay. (The 1.0 release was this month)

I think this is almost ready to go!

murata-yu committed:
Update doc/whats_new/v1.0.rst

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

murata-yu committed:
fixed to addressed until 1.3

murata-yu committed:
fixed to include the same code path

murata-yu committed:
Merge branch 'my_feature' of https://github.com/murata-yu/scikit-learn into my_feature

murata-yu committed:
Merge branch 'main' into my_feature

thomasjpfan committed:
DOC Update doc

thomasjpfan committed:
STY Run black formatting

thomasjpfan reviewed (approved):
I pushed a commit to clean up a comment and some formatting. Otherwise LGTM!

@glemaitre This changed a little bit since your last approval. Maybe this needs another look?

hellojinwoo commented:
> I pushed a commit to clean up a comment and some formatting. Otherwise LGTM!
> 
> @glemaitre This changed a little bit since your last approval. Maybe this needs another look?

Since I would love to use this as soon as possible, I am curious to know when it will be included in the `Scikit-learn 1.1.dev0 (dev) `.
Moreover, will it be included in the 1.1 stable version?

thomasjpfan committed:
TST Fix warning

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pr/20145

thomasjpfan commented:
> Since I would love to use this as soon as possible, I am curious to know when it will be included in the Scikit-learn 1.1.dev0 (dev) .

Once this is merged it will be a part of the nightly build within 24 hours.

>  Moreover, will it be included in the 1.1 stable version?

This PR is very likely to make it into 1.1 stable.

murata-yu committed:
Update sklearn/feature_selection/_sequential.py

Co-authored-by: bmreiniger <40676050+bmreiniger@users.noreply.github.com>

glemaitre committed:
update whats new

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/murata-yu/20145

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pr/20145

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into my_feature

thomasjpfan committed:
DOC Resolve merge issue

cmarmo commented:
Hello @murata-yu , thanks for your work and your patience!
You already have two approvals, do you mind synchronizing with upstream? Then perhaps @agramfort might be available to merge? Thanks!

murata-yu commented:
Thank you @cmarmo .

of course!
There is no problem with being merged.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pr/20145

thomasjpfan commented:
@murata-yu I just pushed a commit to sync with `main`. Once the tests pass, I will merge.

thomasjpfan committed:
DOC Fixes warning
